### PR TITLE
Support Update and Delete for projects

### LINKED ISSUE #215: Projects owned by team are not unique
/v1/projects returns projects like:

```
{
  id: '1a24',
  type: 'project',
  externalIdType: 'project:github_team',
  externalId: 'dev'
}
```

Using a team as the owner of a project doesn't work right, because their names are only unique within an organization, not across them.  (e.g. both rancherio and rancherlabs have a team named "dev").

Need to store, accept and return team id (or org name + team name) instead.

### PR DESCRIPTION
**Pull request by sidharthamani**
@ibuildthecloud

### CODE REVIEW

**Comment by ibuildthecloud** [comment]
I'm not going to merge until i see a test for this issue.

**Comment by sidharthamani** [comment]
@ibuildthecloud - i've added validation tests for this

File: code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/auth/dao/impl/AuthDaoImpl.java [hunk 1]
```diff
@@ -134,6 +140,38 @@ public void updateAccount(Account account, String name, String kind, String exte
             throw new RuntimeException("UpdateAccount failed.");
         }
     }
+    
+    @Override
+    public void updateAccount(String id, Map<String, Object> account) {
```
Comment (ibuildthecloud) [inline code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/auth/dao/impl/AuthDaoImpl.java]: Just use objectmanager.setFields() in resource manager

File: code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/auth/github/ProjectResourceManager.java [hunk 2]
```diff
@@ -58,6 +60,17 @@ public Object listInternal(SchemaFactory schemaFactory, String type, Map<Object,
         if (StringUtils.isEmpty(token)) {
             return null;
         }
+
+        String id = (String) criteria.get("id");
+        if (StringUtils.isNotEmpty(id)) {
+            Account account = authDao.getAccountById(Long.parseLong(id));
```
Comment (ibuildthecloud) [inline code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/auth/github/ProjectResourceManager.java]: Are you checking if the account is really accessible by the user?

File: code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/auth/github/ProjectResourceManager.java [hunk 3]
```diff
@@ -156,6 +174,48 @@ private String getExternalId(String externalId, String externalIdType, String jw
...
+            throw new ClientVisibleException(ResponseCodes.METHOD_NOT_ALLOWED);
+        }
+        authDao.deleteAccount((Account) obj);
+        GithubUtils.ReverseMappings reverseMappings = githubUtils.validateAndFetchReverseMappings(token);
+        return ImmutableList.of(transformProject(authDao.getAccountById(Long.parseLong(id)), reverseMappings));
+    }
+
+    @Override
+    protected Object updateInternal(String type, String id, Object obj, ApiRequest apiRequest) {
+        HttpServletRequest request = apiRequest.getServletContext().getRequest();
```
Comment (ibuildthecloud) [inline code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/auth/github/ProjectResourceManager.java]: Cleanup boilerplate code

File: code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/auth/github/ProjectResourceManager.java [hunk 4]
```diff
@@ -156,6 +174,48 @@ private String getExternalId(String externalId, String externalIdType, String jw
...
+        String token = request.getHeader(AUTH);
+        if (StringUtils.isEmpty(token)) {
+            throw new ClientVisibleException(ResponseCodes.METHOD_NOT_ALLOWED);
+        }
+        Map<String, Object> account = CollectionUtils.toMap(apiRequest.getRequestObject());
+        if (StringUtils.isNotEmpty((String) account.get(EXTERNAL_ID_TYPE))) {
+            String externalId = getExternalId((String) account.get(EXTERNAL_ID), (String) account.get(EXTERNAL_ID_TYPE), token);
+            if (StringUtils.isEmpty(externalId)) {
+                return null;
```
Comment (ibuildthecloud) [inline code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/auth/github/ProjectResourceManager.java]: Throw forbidden exception.

File: code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/auth/dao/impl/AuthDaoImpl.java [hunk 5]
```diff
@@ -99,6 +100,11 @@ public Account createAccount(String name, String kind, String externalId, String
     }
 
     @Override
+    public void deleteAccount(Account account) {
+        objectManager.delete(account);
```
Comment (ibuildthecloud) [inline code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/auth/dao/impl/AuthDaoImpl.java]: Do this from ResourceManager, not DAO.  Remove delete from DAO

```java
scheduleProcess(StandardProcess.REMOVE, obj, null);
return objectManager.reload(obj);
```

Comment (ibuildthecloud) [inline code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/auth/dao/impl/AuthDaoImpl.java]: In account *purge* you need a process handler (extend AbstractDefaultProcessHandler and call it AccountPurge) that does:

Deactivate then remove all hosts.

Create this as a generic "AccountCleanup" handler.  Which will continue to add more account cleanup operations to.

**Comment by ibuildthecloud** [comment]
Add validation test for get by id and delete.

Add integration test for account delete.  Create account, add host to account, do delete, wait for "removed", call account.purge(), wait for purged.  Then assert all hosts are "removed".

Cherry pick https://github.com/ibuildthecloud/cattle/commit/d606b8aec08bc9bbba1871932a97dea1b02ef376 which has create_sim_context to add a host.  Use https://github.com/ibuildthecloud/cattle/commit/d606b8aec08bc9bbba1871932a97dea1b02ef376#diff-cd832280501816aefc90a689f4015d90R187 as an example how to use.

**Comment by ibuildthecloud** [comment]
Do not pass Go. Do not collect $200.  Tests failed (flake8 failed too).

File: code/iaas/api-logic/pom.xml [hunk 6]
```diff
@@ -58,5 +58,10 @@
             <groupId>org.apache.httpcomponents</groupId>
             <artifactId>fluent-hc</artifactId>
         </dependency>
+        <dependency>
+        	<groupId>io.cattle</groupId>
```
Comment (ibuildthecloud) [inline code/iaas/api-logic/pom.xml]: four spaces, no tabs.

File: code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/auth/github/ProjectResourceManager.java [hunk 7]
```diff
@@ -156,6 +193,68 @@ private String getExternalId(String externalId, String externalIdType, String jw
...
+        }
+        String token = getTokenFromRequest(apiRequest);
+        objectProcessManager.executeStandardProcess(StandardProcess.DEACTIVATE, obj,
+                ProcessUtils.chainInData(new HashMap<String, Object>(), AccountConstants.ACCOUNT_DEACTIVATE, AccountConstants.ACCOUNT_REMOVE));
+        Account deletedAccount = (Account) objectManager.reload(obj);
+        GithubUtils.ReverseMappings reverseMappings = githubUtils.validateAndFetchReverseMappings(token);
+        AccountPolicy policy = (AccountPolicy) ApiContext.getContext().getPolicy();
+        Account modifiedAccount = transformProject(deletedAccount, reverseMappings);
+        policy.grantObjectAccess(modifiedAccount);
+        return ImmutableList.of(modifiedAccount);
```
Comment (ibuildthecloud) [inline code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/auth/github/ProjectResourceManager.java]: Avoid trickling guava everywhere.  There's not a great advantage of this over the standard Arrays.asList(...).  I know it's not immutable, but that's not that big of deal.  I'll be honest, I'm not a big fan of Guava, so if you use it it better provide some real value.  The reason it's on the classpath is just for listenablefutures, in general I prefer apache utils.

**Comment by ibuildthecloud** [comment]
@sidharthamani just fix the test and i'll merge

**Pull request merged**